File: src/mlx_omnia/app/api/daemon.py
"""The engine process, and whether this window is the one that started it.

The window discovers a daemon on /admin/health, starts one when nobody answers, and owns
only the one it started: quitting takes that daemon with it, and Settings' Stop/Restart
refuse on any other with the reason.
"""

from __future__ import annotations

import asyncio
import atexit
import contextlib
import os
import pathlib
import signal
import sys

import flet as ft
import httpx

from mlx_omnia import paths
from mlx_omnia.app.api.http import BASE

# src/mlx_omnia/app/api/daemon.py — the checkout is four up, and it is where `uv run` has
# to be pointed when the console script is not on this path.
ROOT = pathlib.Path(os.environ.get("OMNIA_PROJECT") or pathlib.Path(__file__).parents[4])

def bundled_python() -> pathlib.Path | None:
    """The engine's own interpreter inside the .app, or None outside one.

    Flet embeds Python as a dylib in the Flutter host and ships no executable, so in the
    packaged window `sys.executable` is the window itself and there is nothing there to
    exec. `mise run dmg` lays a CPython beside it with the engine installed into it.

    Nothing absolute is stored, which is what lets the bundle be dragged into /Applications
    — the same reason the console scripts in `engine/bin` are unusable here, since their
    shebangs hold the path they were installed under.
    """
    app = paths.bundle()
    if app is None:
        return None
    python = app / "Contents" / "Resources" / "engine" / "bin" / "python3"
    return python if python.exists() else None

FOREIGN = "The engine was started outside the app. Stop it where it was started."


@ft.observable
class Daemon:
    def __init__(self) -> None:
        self.process: asyncio.subprocess.Process | None = None
        # Outlives a failed start, so a daemon that never comes up is not respawned on
        # every probe.
        self.claimed = False

    @property
    def owned(self) -> bool:
        return self.process is not None

    async def boot(self) -> None:
        """Start one if nobody answers. Never blocks the window: the rail's dot is what
        reports a daemon that did not come up."""
        if self.claimed or await up():
            return
        self.claimed = True
        self.process = await _spawn()

    async def stop(self) -> None:
        """SIGTERM and wait: uvicorn shuts down and the port is free for a respawn. No
        SIGKILL fallback — a daemon that ignores SIGTERM is reported, not escalated."""
        child = self.process
        if child is None:
            raise RuntimeError(
                "the daemon was not started by this app; stop it where it was started"
            )
        self.process, self.claimed = None, False
        child.send_signal(signal.SIGTERM)
        try:
            await asyncio.wait_for(child.wait(), 10)
        except TimeoutError:
            self.process, self.claimed = child, True
            raise RuntimeError("the daemon did not exit within 10s of SIGTERM") from None

    async def restart(self) -> None:
        if self.owned:
            await self.stop()
        elif await up():
            raise RuntimeError(
                "the daemon was not started by this app; restart it where it was started"
            )
        self.claimed = True
        self.process = await _spawn()


def _term(pid: int) -> None:
    with contextlib.suppress(OSError):
        os.kill(pid, signal.SIGTERM)


async def up() -> bool:
    try:
        async with httpx.AsyncClient(base_url=BASE, timeout=2) as client:
            await client.get("/admin/health")
    except httpx.HTTPError:
        return False
    return True


def _command() -> list[str]:
    """How to start the engine, in the three places the window runs from."""
    if (bundled := bundled_python()) is not None:
        return [str(bundled), "-m", "mlx_omnia.server.main"]
    beside = pathlib.Path(sys.executable).parent / "omnia-server"
    if beside.exists():
        return [str(beside)]
    return ["uv", "run", "--project", str(ROOT), "omnia-server"]


def child_environment() -> dict[str, str]:
    """The child's environment, with this interpreter's own settings taken out of it.

    serious_python points `PYTHONHOME` and `PYTHONPATH` at the window's embedded runtime,
    and a child inherits them: the engine's interpreter would then read the window's stdlib
    and the window's site-packages, where the server's dependencies are deliberately not.
    Inherited, it does not even reach an import error — `Py_Initialize` fails and the
    process dies with no Python frame to blame.
    """
    return {
        key: value
        for key, value in os.environ.items()
        if not key.startswith("PYTHON") and key != "SERIOUS_PYTHON_APP"
    }


async def _spawn() -> asyncio.subprocess.Process:
    command = _command()
    bundled = bundled_python() is not None
    # The checkout is where `uv run` has to start; the bundle has no checkout, and `ROOT`
    # there is whatever lies four levels above a temp directory. The engine needs no working
    # directory of its own — the model cache is addressed absolutely through the hub.
    cwd = None if bundled else ROOT
    # The daemon's half of "quit is a stop": it watches this pid and shuts itself down
    # when it goes, which is the only path that survives a force quit.
    # Nobody is watching this child's terminal — in the bundle there is not one. Truncated
    # at each start, because the run whose lines answer "why is it not up" is the one that
    # has just failed, and uvicorn writes an access line per request.
    handle = paths.daemon_log().open("wb")
    try:
        process = await asyncio.create_subprocess_exec(
            *command,
            "--parent-pid",
            str(os.getpid()),
            cwd=cwd,
            env=child_environment() if bundled else None,
            stdout=handle,
            stderr=asyncio.subprocess.STDOUT,
        )
    finally:
        # The child holds its own descriptor now.
        handle.close()
    # Quit is a stop, immediately. The --parent-pid watchdog is what covers the paths
    # this never runs on.
    atexit.register(_term, process.pid)
    return process

File: app/Sources/OmniaPanel/Daemon.swift
// The engine process, and whether this panel is the one that started it.
//
// The panel discovers a daemon on /admin/health, starts one when nobody answers, and owns
// only the one it started: quitting takes that daemon with it, and Stop/Restart refuse on
// any other with the reason. Same rule as the window's `app/api/daemon.py`, and the same
// `--parent-pid` handshake — the daemon watches this pid and shuts itself down when it
// goes, which is the only path that survives a force quit.

import Foundation
import Observation

@MainActor
@Observable
final class Daemon {
    private var process: Process?

    var owned: Bool { process?.isRunning == true }

    static let foreign = "The engine was started outside the panel. Stop it where it was started."

    /// The checkout this panel was built inside, which is where `uv run` has to start and
    /// where the venv's console scripts live. `OMNIA_PROJECT` overrides it, exactly as the
    /// window's module does.
    static let root: URL? = {
        if let named = ProcessInfo.processInfo.environment["OMNIA_PROJECT"] {
            return URL(fileURLWithPath: named)
        }
        var here = URL(fileURLWithPath: CommandLine.arguments[0]).resolvingSymlinksInPath()
        for _ in 0..<12 {
            here.deleteLastPathComponent()
            if here.path == "/" { break }
            let marker = here.appendingPathComponent("pyproject.toml")
            if FileManager.default.fileExists(atPath: marker.path),
               let text = try? String(contentsOf: marker, encoding: .utf8),
               text.contains("mlx-omnia") {
                return here
            }
        }
        return nil
    }()

    /// The interpreter shipped inside this .app, or nil outside one. `mise run dmg` lays a
    /// CPython with the engine installed into it under Contents/Resources/engine; a panel
    /// built by `build.sh` alone has none, and falls through to the checkout.
    ///
    /// The console scripts are not used even when they are there: their shebang holds the
    /// path they were installed under, which is wrong the moment the .app is dragged into
    /// /Applications. `-m` is addressed relative to the interpreter that runs it.
    static var bundled: URL? {
        guard let resources = Bundle.main.resourceURL else { return nil }
        let python = resources.appendingPathComponent("engine/bin/python3")
        return FileManager.default.fileExists(atPath: python.path) ? python : nil
    }

    /// How to start one of Omnia's console scripts, in the three places the panel runs from:
    /// the interpreter inside the bundle, the checkout's venv, or `uv` when there is no venv
    /// to point at. The module is named for the bundle's sake — a script is a path there,
    /// and `-m` is not.
    static func command(_ script: String, module: String) -> (URL, [String])? {
        if let bundled { return (bundled, ["-m", module]) }
        guard let root else { return nil }
        let beside = root.appendingPathComponent(".venv/bin/\(script)")
        if FileManager.default.fileExists(atPath: beside.path) { return (beside, []) }
        let places = ["/opt/homebrew/bin/uv", "/usr/local/bin/uv",
                      NSHomeDirectory() + "/.local/bin/uv"]
        guard let uv = places.first(where: { FileManager.default.fileExists(atPath: $0) })
        else { return nil }
        return (URL(fileURLWithPath: uv), ["run", "--project", root.path, script])
    }

    static var logs: URL {
        let directory = URL(fileURLWithPath: NSHomeDirectory())
            .appendingPathComponent("Library/Logs/mlx-omnia")
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory
    }

    static func up() async -> Bool {
        var request = URLRequest(url: URL(string: Client.base.absoluteString + "/admin/health")!)
        request.timeoutInterval = 1.5
        guard let (_, answer) = try? await URLSession.shared.data(for: request),
              let answer = answer as? HTTPURLResponse
        else { return false }
        return (200..<300).contains(answer.statusCode)
    }

    /// Start one if nobody answers. Raises rather than swallowing, because the two callers
    /// — the launch and the button on the down card — both have somewhere to put the
    /// reason.
    func boot() async throws {
        if owned { return }
        if await Self.up() { return }
        try spawn()
    }

    private func spawn() throws {
        guard let (executable, arguments) =
            Self.command("omnia-server", module: "mlx_omnia.server.main")
        else {
            throw Refused(detail: "no engine in this bundle and no checkout to start one from")
        }
        // Truncated at each start, because the run whose lines answer "why is it not up"
        // is the one that has just failed, and uvicorn writes an access line per request.
        let log = Self.logs.appendingPathComponent("daemon.log")
        let handle = try LogTail.openForWriting(log)

        let child = Process()
        child.executableURL = executable
        child.arguments = arguments + ["--parent-pid", String(ProcessInfo.processInfo.processIdentifier)]
        child.currentDirectoryURL = Self.root
        child.standardOutput = handle
        child.standardError = handle
        try child.run()
        process = child
    }

    /// SIGTERM and wait: uvicorn shuts down and the port is free for a respawn. No SIGKILL
    /// fallback — a daemon that ignores SIGTERM is reported, not escalated.
    func stop() async throws {
        guard let child = process, child.isRunning else {
            throw Refused(detail: Self.foreign)
        }
        process = nil
        child.terminate()
        await withCheckedContinuation { (settled: CheckedContinuation<Void, Never>) in
            DispatchQueue.global().async {
                child.waitUntilExit()
                settled.resume()
            }
        }
    }

    func restart() async throws {
        if owned {
            try await stop()
        } else if await Self.up() {
            throw Refused(detail: Self.foreign)
        }
        try await boot()
    }

    /// Quitting takes the daemon this panel started with it.
    func standDown() {
        process?.terminate()
    }

}
